keps/sig-apps/4563-evacuation-api/README.md

```diff
@@ -0,0 +1,1616 @@
+will proceed with the eviction. To support this, the eviction could only be allowed only at this
+point. That is, if there is a matching Evacuation object and the eviction conditions are met
+according to the [Eviction](#eviction) section.
+
+Eviction of these annotated pods would act as a PDB with a fallback to eviction if things do not go
+according to plan.
+
+This would break the eviction assumption, that eviction can proceed quickly when no PDB exists or is
+non-blocking, but it would greatly increase the safety and availability for applications.
+
+This would reduce the need to override the eviction endpoint with validating admission webhooks, as
+we can see in many projects today. They would only need a single controller responding to
+Evacuation objects.
+
+
+### Evacuation Process
+
+The following diagrams describe what the evacuation process will look like for each actor:
+
+![evacuation-process](evacuation.svg)
+
+
+### Follow-up Design Details for Kubernetes Workloads
+
+Kubernetes Workloads should be made aware of the Evacuation API to properly support the evacuation
+process.
+
+- ReplicaSet and ReplicationController partial support is considered. The full evacuation logic
+  should be implemented by a higher level workload such as Deployments. ReplicaSets are intended to
+  be a simple primitive without any sophisticated scaling logic.
+- Deployment support is considered.
+- StatefulSet support is considered.
+- DaemonSet support is considered, but will be explored as a part of the 
+  [Declarative Node Maintenance KEP](https://github.com/kubernetes/enhancements/pull/4213), as its
+  pods should not be disrupted/evacuated during normal operation.
+- Jobs/CronJobs are not yet considered, because jobs should not normally run critical workloads
+  that require evacuation and should therefore leave the node quickly and not block its drain.
+  If required, custom Evacuators can be implemented to support custom evacuation for specialized
+  types of jobs.
+
+#### ReplicaSet Controller
+
+Active response to Evacuation objects cannot be implemented without first analyzing a pod's owner
+and its capabilities. For example, not all deployment strategy configurations can simply be
+evacuated. This is also true for other owners. Therefore, it does not make sense to impose such a
+responsibility on the replication controller (e.g., by introducing new fields). Instead,
+deployments and other controllers should advertise the evacuation capabilities on ReplicaSet pods
+and implement/perform the evacuation.
+
+To facilitate the evacuation of high-level workloads and to take advantage of the Evacuation
+feature passively, the replicaset controller can prefer the deletion of pods with corresponding
+Evacuation objects during a ReplicaSet scale down.
+
+#### Deployment Controller
+
+The deployment controller must first decide whether a particular Deployment supports the evacuation
```

Comment:
Do we need to consider integration with HPA? During an evacuation, if HPA scales down a Deployment, do we just follow the existing ReplicaSet's scale-down behavior to delete those Pods that need to be reduced, aside from the Pods scheduled for termination in the Evacuation API?

Reply:
Regardless of the Deployment strategy, if the Deployment is scaled down during an evacuation, the ReplicaSet is scaled down and the pods with matching Evacuations can be terminated immediately by the replica set controller.

There are ways for the HPA to inject itself into the evacuation process. For example it could become the Evacuation Instigator and select the pods that should be scaled down by creating Evacuations for supported workloads. This is not a feature that HPA offers today, and I am not sure if it should.  I think it is better to leave it out of this KEP.

The Deployments and ReplicaSets should work for both the manual and automatic scaling.

Reply:
@toVersus I have added a new followup for the HPA use case (see [HorizontalPodAutoscaler](https://github.com/kubernetes/enhancements/blob/6d3a013d93e0cc84d82a7394fee3ba1913acc88f/keps/sig-apps/4563-evacuation-api/README.md#horizontalpodautoscaler)).

- It should be possible for the HPA to become the evacuator of the workload, if Evacuations are observed. 
- I also roughly outline how the evacuations conflicts should be resolved.

Reply:
Thank you for adding the HPA controller section in the Follow-up design!

> Upon detecting an evacuation of pods under the HPA workload, the HPA would increase the number of replicas by the number of evacuations.

Just to confirm, the behavior described here refers to when the HPA controller becomes the Evacuator (i.e., when evacuation priority annotations are added to the HPA object), right? I agree that, to avoid conflicts with other pod controllers, I agree that the HPA controller does not perform evacuation by default.

Although this is something we can decide during future discussions about pod controller's Evacuation API support, another concern is whether the HPA controller will modify the replicas of Deployments, etc., during evacuation. If the pod controller increases the replicas to account for maxSurge and then the HPA controller modifies the replicas, there could be side effects. It seems simpler for the HPA controller not to perform scale-in or scale-out during evacuation, but if the evacuation takes a long time, not being able to scale in or scale out for a long period of time could itself be problematic.

Reply:
Yes, it should become an evacuator and add annotations to pods. This is up for disscusion if the evacuation should be by default or not. I do not see a downside, by adding it as a lower priority evacuator.

The priorities and scaling operations depend on the workload. But it should never happen that both the deployment controller and the HPA increase replicas/pods at the same time.